numlib/changeuidgid.c

```diff
@@ -78,27 +79,31 @@ uid_t libmail_getuid(const char *uname, gid_t *pw_gid)
 		bufsize = 16384;        /* Should be more than enough */
 	}
 
-	buf = malloc(bufsize);
-	if (buf == NULL)
-	{
-		perror("malloc");
-		exit(1);
-	}
-
-
-	errno=ENOENT;
+	do {
+		buf = malloc(bufsize);
+		if (buf == NULL)
+		{
+			perror("malloc");
+			exit(1);
+		}
 
-	getpwnam_r(p, &pwbuf, buf, bufsize, &pw);
+		s = getpwnam_r(p, &pwbuf, buf, bufsize, &pw);
+		if (s == ERANGE) {
+			free(buf);
+			bufsize += 1024;
+		}
+	} while (s == ERANGE && bufsize <= 65536);
```

Comment:
I believe this is not the case. That is why I moved the second free behind the 'if (pw == 0)' part. In the error case that part will always get executed, and the function will end with exit. So the second free(buf) will only get executed in the non-error case.

Reply:
Ah yes. I missed that. I don't think this will double free.

Is that the right error handling though? `getpwnam_r` can return success with `pw` set to `NULL`.
> If no matching password record was found, these functions return 0 and store NULL in *result.

I think that'll give a strange error message in those cases. It will be something like `getpwnam_r: Undefined error: 0` (that's what `perror` gives me with `errno` set to 0 on macOS, it's probably slightly different for glibc).